clients/src/main/java/org/apache/kafka/common/requests/RequestHeader.java

```diff
@@ -75,7 +77,17 @@ public void write(ByteBuffer buffer, ObjectSerializationCache serializationCache
     }
 
     public int size(ObjectSerializationCache serializationCache) {
-        return data.size(serializationCache, headerVersion);
+        if (this.size == SIZE_NOT_INITIALIZED) {
```

Comment:
I think this method does not need to be public anymore, callers from other packages use the new overload. 
Also on the golden path `size() -> size(ObjectSerializationCache serializationCache)` we do the equality check twice. If we inline it, does it make a difference?

Reply:
1. I changed the access modifier for this method to default. It is used by tests hence, I am currently not changing to private.

2. I have made similar changes to RespondeHeader as well to prevent incorrect future usage of the size() method over there that would cause similar performance problems.

3. Added JavaDocs primarily meant for contributors to ensure that they use the correct intended method.

4. In this refactor, the size(ObjectSerializationCache) method calculates the size every time (instead of using the cached value). This should remove the double equality checks.